### PR TITLE
Implement IOperation for suppression operator

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/31582

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -6608,6 +6608,17 @@ public override IOperation VisitRangeOperation(IRangeOperation operation, int? a
             return new RangeOperation(operation.IsLifted, operation.IsImplicit, semanticModel: null, operation.Syntax, operation.Type, visitedLeftOperand, visitedRightOperand, operation.Method);
         }
 
+        public override IOperation VisitSuppressNullableWarningOperation(ISuppressNullableWarningOperation operation, int? argument)
+        {
+            IOperation visitedExpression = null;
```
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Null check is unnecessary.

**Comment by 333fred** [comment]
Done review pass (commit 2)

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs [hunk 2]
```diff
@@ -9183,6 +9183,76 @@ public override IOperation RightOperand
...
+        }
+
+        public abstract IOperation Expression { get; }
+
+        public sealed override IEnumerable<IOperation> Children
+        {
+            get
+            {
+                IOperation expression = Expression;
+                if (expression != null)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs]: >if (expression != null) [](start = 16, length = 23)

Can this ever be false?

File: src/Compilers/Core/Portable/Operations/OperationNodes.cs [hunk 3]
```diff
@@ -9183,6 +9183,76 @@ public override IOperation RightOperand
         }
     }
 
+    internal abstract class BaseSuppressNullableWarningOperation : Operation, ISuppressNullableWarningOperation
+    {
+        protected BaseSuppressNullableWarningOperation(bool isImplicit, SemanticModel semanticModel, SyntaxNode syntax, ITypeSymbol type) :
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/OperationNodes.cs]: >bool isImplicit [](start = 55, length = 15)

We usually keep ```isImplicit``` parameter last

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 4]
```diff
@@ -6608,6 +6608,17 @@ public override IOperation VisitRangeOperation(IRangeOperation operation, int? a
...
+            if (!(operation.Expression is null))
+            {
+                visitedExpression = Visit(operation.Expression);
+            }
+
+            return new SuppressNullableWarningOperation(operation.IsImplicit, semanticModel: null, operation.Syntax, operation.Type, visitedExpression);
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >operation.IsImplicit [](start = 56, length = 20)

```IsImplicit(operation)```

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: It looks like two methods above have the same problem. Please fix them too.

---
In reply to: [244069749](https://github.com/dotnet/roslyn/pull/32019#discussion_r244069749) [](ancestors = 244069749)

File: src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs [hunk 5]
```diff
@@ -1332,6 +1332,22 @@ public override void VisitRangeOperation(IRangeOperation operation)
...
 
+        public override void VisitSuppressNullableWarningOperation(ISuppressNullableWarningOperation operation)
+        {
+            Assert.Equal(OperationKind.SuppressNullableWarning, operation.Kind);
+
+            IOperation[] children = operation.Children.ToArray();
+
+            int index = 0;
+
+            if (operation.Expression != null)
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs]: Yes please simply unconditionally assert:
```
Assert.Same(operation.Expression, operation.Children.Single())
```

---
In reply to: [244038738](https://github.com/dotnet/roslyn/pull/32019#discussion_r244038738) [](ancestors = 244038738)

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests.cs [hunk 6]
```diff
@@ -13,6 +13,60 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
...
+        [Fact]
+        public void SuppressNullableWarningOperation()
+        {
+            var comp = CreateCompilation(@"
+class C
+{
+#nullable enable
+    void M(string? x)
+    /*<bind>*/{
+        x!.ToString();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests.cs]: >x [](start = 8, length = 1)

We should have a simple flow graph test for a scenario with control flow inside the underlying expression.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests.cs [hunk 7]
```diff
@@ -13,6 +13,60 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
...
+    }/*</bind>*/
+}");
+            comp.VerifyDiagnostics();
+
+            var m = comp.SyntaxTrees.Single().GetRoot().DescendantNodes().OfType<BlockSyntax>().Single();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests.cs]: >var m = comp.SyntaxTrees.Single().GetRoot().DescendantNodes().OfType<BlockSyntax>().Single(); [](start = 12, length = 93)

Is this local used?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

**Comment by jcouv** [comment]
>     internal abstract class LazyFromEndIndexOperation : BaseFromEndIndexOperation

@agocke Not related to this PR: I remember that LDM settled on a name for the `^` operator. Do we need to update our public API (syntax, IOperation, ...)?

---
Refers to: src/Compilers/Core/Portable/Operations/OperationNodes.cs:9059 in aafd4e6. [](commit_id = aafd4e6802a420ac42689907a9ee1a00ba25421b, deletion_comment = False)

**Comment by jaredpar** [comment]
approved to merge (pending sign offs)

**Pull request merged**